hoodie-client/src/main/java/com/uber/hoodie/func/BufferedIterator.java

```diff
@@ -0,0 +1,224 @@
+  }
+
+  private void insertRecord(T t) throws Exception {
+    adjustBufferSize(t);
+    rateLimiter.acquire();
+    // New Hoodie record is created to ensure that all the computation happens in reader thread.
+    HoodieRecordPayload payload = t.getData();
+    HoodieRecord newRecord = new HoodieRecord<>(t.getKey(),
+      new BufferedInsertPayload(payload.getInsertValue(schema), payload.getMetadata()));
+    buffer.put(new BufferedIteratorRecord(Optional.of(newRecord)));
+  }
+
+  private void readNextRecord() {
+    if (readCounter.incrementAndGet() % 1024 == 0) {
+      logger.debug("oldRateLimit:" + currentRateLimit + ":readCounter:" + readCounter.get()
+        + ":currentSize:" + buffer.size() + ":avgRecordSize:" + avgRecordSizeOfSampledRecords);
+    }
+    rateLimiter.release();
+    BufferedIteratorRecord<T> newRecord;
+    while (true) {
+      try {
+        throwExceptionIfFailed();
+        newRecord = buffer.poll(5, TimeUnit.SECONDS);
+        if (newRecord != null) {
+          break;
+        }
+      } catch (InterruptedException e) {
+        logger.error("error reading records from BufferedIterator", e);
+        throw new HoodieException(e);
+      }
+    }
+    if (newRecord.record.isPresent()) {
+      this.nextRecord = newRecord.record.get();
+    } else {
+      // We are done reading all the records from internal iterator.
+      this.isDone.set(true);
+      this.nextRecord = null;
+    }
+  }
+
+  public void startRecordReader() {
+    try {
+      while (internalIterator.hasNext()) {
+        // We need to stop reading if parquet writer thread has failed to write records.
+        throwExceptionIfFailed();
+        insertRecord(internalIterator.next());
+      }
+      // done reading records notifying reader.
+      buffer.put(new BufferedIteratorRecord<>(Optional.absent()));
+    } catch (Exception e) {
+      logger.error("error reading records", e);
+      throw new HoodieException(e);
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (this.nextRecord == null && !this.isDone.get()) {
+      readNextRecord();
+    }
+    return !this.isDone.get();
+  }
+
+  @Override
+  public T next() {
+    if (this.nextRecord == null && !this.isDone.get()) {
+      readNextRecord();
+    }
+    Preconditions.checkState(this.nextRecord != null);
+    final T ret = this.nextRecord;
+    this.nextRecord = null;
+    return ret;
+  }
+
+  private void throwExceptionIfFailed() {
+    if (this.hasFailed.get() != null) {
+      throw new HoodieException("operation has failed", this.hasFailed.get());
+    }
+  }
+
+  public void markAsFailed(Exception e) {
+    this.hasFailed.set(e);
+    this.rateLimiter.release(RECORD_CACHING_LIMIT + 1);
+  }
+
+  private static final class BufferedIteratorRecord<T> {
+
+    private Optional<T> record;
+
+    BufferedIteratorRecord(Optional<T> record) {
+      this.record = record;
+    }
+  }
+
+  private static final class BufferedInsertPayload implements HoodieRecordPayload {
```

Comment:
same here.. can we avoid this class, by narrowing/adjusting the generic type def in the class ?

Reply:
sure. I wanted this to offload computing to reader thread. Let me introduce another api in HoodieRecord -> prepareInsertValue(). Let me know what you think.